Add ability for a user to set mqtt's `quiescentTimeout` for forceable shutdowns

cppwfs opened the pull request:
* Mqttv5 updated to  also support this feature
* Fixes: https://github.com/spring-projects/spring-integration/issues/10095

cppwfs committed:
Add ability for a user to set mqtt's `quiescentTimeout` for forceable shutdowns

Mqttv5 updated to  also support this feature
Fixes: https://github.com/spring-projects/spring-integration/issues/10095

artembilan commented on spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/AbstractMqttMessageDrivenChannelAdapter.java:
```diff
@@ -59,6 +60,8 @@
 public abstract class AbstractMqttMessageDrivenChannelAdapter<T, C> extends MessageProducerSupport
 		implements ApplicationEventPublisherAware, ClientManager.ConnectCallback {
 
+	public static final Long QUIESCENT_TIMEOUT = 30_000L;
```
Move this into a `ClientManager` alongside with the `DISCONNECT_COMPLETION_TIMEOUT`.

artembilan commented on spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/AbstractMqttMessageDrivenChannelAdapter.java:
```diff
@@ -199,6 +204,20 @@ protected long getDisconnectCompletionTimeout() {
 
+	/**
+	 * Set the quiescentTimeout timeout when disconnecting.
+	 * Default is 30,000 milliseconds.
```
```suggestion
	 * Default {@value ClientManager#QUIESCENT_TIMEOUT} milliseconds.
```

artembilan commented on spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/AbstractMqttMessageDrivenChannelAdapter.java:
```diff
@@ -199,6 +204,20 @@ protected long getDisconnectCompletionTimeout() {
+		this.quiescentTimeout = quiescentTimeout;
+	}
+
+	public long getQuiescentTimeout() {
```
``
```suggestion
	protected long getQuiescentTimeout() {
```

artembilan commented on spring-integration-mqtt/src/test/java/org/springframework/integration/mqtt/BackToBackAdapterTests.java:
```diff
@@ -123,6 +128,33 @@ public void testSingleTopic() {
+		QueueChannel outputChannel = new QueueChannel();
+		inbound.setOutputChannel(outputChannel);
+		inbound.setTaskScheduler(taskScheduler);
+		inbound.setQuiescentTimeout(QUIESCENT_TIMEOUT);
```
Let's apply it to all existing tests instead!

artembilan commented on spring-integration-mqtt/src/test/java/org/springframework/integration/mqtt/Mqttv5BackToBackTests.java:
```diff
@@ -153,6 +154,23 @@ public void testSharedTopicMqttv5Interaction() {
+	void testSharedTopicMqttv5InteractionQuiescentTimeout() {
+		this.mqttv5MessageDrivenChannelAdapter.addTopic("$share/group/testTopicq");
+		this.mqttv5MessageDrivenChannelAdapter.setQuiescentTimeout(2000);
+		this.mqttv5MessageDrivenChannelAdapter.setDisconnectCompletionTimeout(2000);
```
Let's apply this to existing tests instead!

artembilan reviewed (changes requested):
Also apply this feature in the `AbstractMqttClientManager` implementations.
Plus mention a new property in the `whats-new.adoc`.

Thanks

cppwfs committed:
Move QUIESCENT_TIMEOUT constant to ClientManager

Update tests to set setDisconnectCompletionTimeout and setQuiescentTimeout to 1L to reduce test runtime
Rebase

artembilan commented on spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/Mqttv5PahoMessageDrivenChannelAdapter.java:
```diff
@@ -296,7 +298,8 @@ protected void doStop() {
 				}
 				if (getClientManager() == null) {
-					this.mqttClient.disconnectForcibly(getDisconnectCompletionTimeout());
+					this.mqttClient.disconnectForcibly(ClientManager.QUIESCENT_TIMEOUT, getDisconnectCompletionTimeout(),
```
Did you mean `getQuiescentTimeout()` as you do in the `MqttPahoMessageDrivenChannelAdapter` for MQTTv3?

cppwfs replied:
DOH!

artembilan commented on spring-integration-mqtt/src/test/java/org/springframework/integration/mqtt/BackToBackAdapterTests.java:
```diff
@@ -396,6 +389,14 @@ public void onApplicationEvent(MqttSubscribedEvent event) {
 
 	}
 
+	private void initializeInboundAdapter(MqttPahoMessageDrivenChannelAdapter inbound, QueueChannel outputChannel) {
```
`static`

artembilan reviewed (changes requested):
See `Mqttv3ClientManager` and `Mqttv5ClientManager` `stop()`.
Such a `quiesceTimeout` has to be implemented over there as well.
Plus `whats-new.adoc` entry.

cppwfs committed:
Mqttv5PahoMessageDrivenChannelAdapter.disconnectForcibly needs to use getQuiescentTimeout for quiescent timeout

BackToBackAdapterTests initializeInboundAdapter needs to be a static  method

cppwfs committed:
Apply quiescentTimeout attribute to AbstractMqttClientManager.

Add the quiescentTimeout in the disconnectForcibly in the Mqttv3ClientManager and Mqttv5ClientManager classes

Add documentation to what's new doc describing the addition of quiescence timeout

artembilan commented on src/reference/antora/modules/ROOT/pages/whats-new.adoc:
```diff
@@ -44,3 +44,8 @@ See xref:redis.adoc#redis-lock-registry[Redis Lock Registry] for more informatio
+[[x7.0-mqtt-changes]]
+=== MQTT Changes
+
+The `disconnectForcibly` method in `MqttPahoMessageDrivenChannelAdapter`, `Mqttv5PahoMessageDrivenChannelAdapter`, `Mqttv3ClientManager`, and `Mqttv5ClientManager` now supports `getQuiescentTimeout` in addition to the existing `disconnectCompletionTimeout`.
```
That’s not quite true. That method had it supported for years in the target client. And in fact that is not a method of our classes. Please , try to be more specific. 
let’s say something like those classes now provide such an option which is propagated down to the respective client API.
Also pay attention to the typo of new property. It is really not `get…` copy/paste artifact I guess 😉
Also link to the target chapter , please.

cppwfs committed:
Update the whats new MQTT announcement

It should be rewritten to state that this is exposing more of the MQTT Client API.

artembilan commented on src/reference/antora/modules/ROOT/pages/whats-new.adoc:
```diff
@@ -48,4 +48,5 @@ Previously deprecated classes in the `spring-integation-hazelcast` module, such
 === MQTT Changes
 
-The `disconnectForcibly` method in `MqttPahoMessageDrivenChannelAdapter`, `Mqttv5PahoMessageDrivenChannelAdapter`, `Mqttv3ClientManager`, and `Mqttv5ClientManager` now supports `getQuiescentTimeout` in addition to the existing `disconnectCompletionTimeout`.
+The `disconnectForcibly` method in `MqttPahoMessageDrivenChannelAdapter`, `Mqttv5PahoMessageDrivenChannelAdapter`, `Mqttv3ClientManager`, and `Mqttv5ClientManager` now expose the option to configure `quiescentTimeout`, which is available through the https://eclipse.dev/paho/files/javadoc/org/eclipse/paho/client/mqttv3/IMqttAsyncClient.html[IMqttAsyncClient] client.
```
Sorry, I still disagree with this sentence.
How about like this:
```
The `AbstractMqttMessageDrivenChannelAdapter` and `ClientManager` implementations now expose a `quiescentTimeout` option which is propagated in their `stop()` method down to the `disconnectForcibly()` API of the MQTT Paho clients.
```
The JavaDoc to MQTT v3 client is not quit right here since we also do the same for MQTT v5 client interaction.

cppwfs replied:
Should we add MQTT V3 and V5 links.   OR remove the V3?

artembilan replied:
Remove altogether.
We may mention something like `See their JavaDocs for more information.`, though.

artembilan commented on src/reference/antora/modules/ROOT/pages/whats-new.adoc:
```diff
@@ -48,4 +48,5 @@ Previously deprecated classes in the `spring-integation-hazelcast` module, such
 
-The `disconnectForcibly` method in `MqttPahoMessageDrivenChannelAdapter`, `Mqttv5PahoMessageDrivenChannelAdapter`, `Mqttv3ClientManager`, and `Mqttv5ClientManager` now supports `getQuiescentTimeout` in addition to the existing `disconnectCompletionTimeout`.
+The `disconnectForcibly` method in `MqttPahoMessageDrivenChannelAdapter`, `Mqttv5PahoMessageDrivenChannelAdapter`, `Mqttv3ClientManager`, and `Mqttv5ClientManager` now expose the option to configure `quiescentTimeout`, which is available through the https://eclipse.dev/paho/files/javadoc/org/eclipse/paho/client/mqttv3/IMqttAsyncClient.html[IMqttAsyncClient] client.
+See https://docs.spring.io/spring-integration/api/org/springframework/integration/mqtt/core/package-summary.html[MQTT Core Java Docs] for more information.
```
This is not right.
We have talked about the target MQTT chapter of this reference manual, but you point to JavaDoc of the package.
It has to be like this:
```
See xref:mqtt.adoc[] for more information.
```

cppwfs replied:
Let's go with the final fix. We can revisit the docs later on, in that our javadocs cover the updates satisfactorly.

cppwfs committed:
Change what's new notes on mqtt changes to be brief

artembilan reviewed (changes requested):
Looks like PR build is failing in the test related to the change.
```
 > Task :spring-integration-mqtt:test
ClientManagerBackToBackTests > testSameV3ClientIdWorksForPubAndSub() FAILED
    java.lang.AssertionError at ClientManagerBackToBackTests.java:132
```
Any thoughts?
Thanks
